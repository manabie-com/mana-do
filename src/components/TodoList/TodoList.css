.app-table {
    border-collapse: collapse;
    width: 100%;
    table-layout: fixed;

}   

.app-table thead tr {
    background-color: #39a1f4;
    color: #fff;

}

.app-table th,
.app-table td {
    padding: 5px 24px;
}

.app-table td {
    white-space:pre-wrap; 
    word-wrap:break-word
}

.app-table tbody tr {
    border-bottom: 1px solid #dddddd;
}

.app-table .th-15-percent {
    width: 15%;
}

.app-table .th-seq-number {
    width: 1%;
}

.app-table .align-center {
    text-align: center;
}

.app-table .align-left {
    text-align: left;
}

.app-table .align-right {
    text-align: right;
}