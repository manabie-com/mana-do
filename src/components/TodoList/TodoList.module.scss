.container {
  min-height: 50vh;
}
