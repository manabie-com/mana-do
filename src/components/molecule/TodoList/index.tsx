import ToDoList from "./TodoList";

export default ToDoList;
