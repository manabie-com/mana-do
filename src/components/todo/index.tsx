import React, { useEffect, useReducer, useRef, useState, useMemo, useCallback } from 'react';
import { RouteComponentProps } from 'react-router-dom';

import reducer, { initialState } from 'root/store/reducer';
import {
  setTodos,
  createTodo,
  deleteTodo,
  toggleAllTodos,
  deleteAllTodos,
  updateTodoStatus,
  updateTodoContent
} from 'root/store/actions';
import Service from 'root/service';
import { TodoStatus } from 'root/models/todo';
import { isTodoCompleted } from 'root/utils';
import TodoPresentation from './presentation';

import './_todo.css'

type EnhanceTodoStatus = TodoStatus | 'ALL';


const TodoComponent = ({ history }: RouteComponentProps) => {
  const [{ todos }, dispatch] = useReducer(reducer, initialState);
  const [showing, setShowing] = useState<EnhanceTodoStatus>('ALL');
  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    (async () => {
      const resp = await Service.getTodos();
      dispatch(setTodos(resp || []));
    })()
  }, [])

  useEffect(() => {
    if (todos.length > 0) {
      const lastItem = document.getElementById(`todo-item-${todos[todos.length - 1].id}`)
      if (lastItem) setTimeout(() => {
        lastItem.scrollIntoView({ block: 'end',  behavior: 'smooth' })
      }, 401) // delay to wait for css animation work first
    }
  }, [todos, showing])

  const onCreateTodo = useCallback(async (e: React.KeyboardEvent<HTMLInputElement>) => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    if (e.key === 'Enter' && inputRef.current) {
      try {
        const resp = await Service.createTodo(inputRef.current.value);
        dispatch(createTodo(resp));
        inputRef.current.value = '';
      } catch (e) {
        if (e.response.status === 401) {
          history.push('/')
        }
      }
    }
  }, [dispatch, history])

  const onUpdateTodoStatus: Function = useCallback(async (e: React.ChangeEvent<HTMLInputElement>, todoId: string) => {
    try {
      await Service.updateTodoStatus(todoId, e.target.checked);
      dispatch(updateTodoStatus(todoId, e.target.checked));
    } catch (error) {

    }
  }, [dispatch])

  const onToggleAllTodo = useCallback(async (e: React.ChangeEvent<HTMLInputElement>) => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    try {
      await Service.toggleAllTodos(e.target.checked)
      dispatch(toggleAllTodos(e.target.checked))
    } catch (error) {

    }
  }, [dispatch])

  const onDeleteAllTodo = useCallback(async () => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    try {
      await Service.deleteAllTodo()
      dispatch(deleteAllTodos());
    } catch (error) {

    }
  }, [dispatch])

  const handleDeleteTodo = useCallback(async (todoId: string) => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    try {
      await Service.deleteTodo(todoId)
      dispatch(deleteTodo(todoId));
    } catch (error) {

    }
  }, [dispatch])

  const handleUpdateTodoContent = useCallback(async (todoId: string, content: string) => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    try {
      await Service.updateTodoContent(todoId, content)
      dispatch(updateTodoContent(todoId, content));
    } catch (error) {

    }
  }, [dispatch])

  const handleShowingAll = useCallback(() => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    setShowing('ALL')
  }, [setShowing])

  const handleShowingActive = useCallback(() => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    setShowing(TodoStatus.ACTIVE)
  }, [setShowing])

  const handleShowingCompleted = useCallback(() => {
    /**
     * reduce rebinding function. Does not rebind when dependencies has not changed
     */
    setShowing(TodoStatus.COMPLETED)
  }, [setShowing])

  const showTodos = useMemo(() => todos.filter((todo) => {
    /**
     * Reduce recalculating showing todos, Does not recalculate showing todos when todos and showing state have not changed
     */
    switch (showing) {
      case TodoStatus.ACTIVE:
        return todo.status === TodoStatus.ACTIVE;
      case TodoStatus.COMPLETED:
        return todo.status === TodoStatus.COMPLETED;
      case 'ALL':
      default:
        return true;
    }
  }), [showing, todos])

  const activeTodos = useMemo(() => todos.reduce(function (accum, todo) {
    /**
     * Reduce recalculating activeTodos, Does not recalculate active todos when todos has not changed
    */
    return isTodoCompleted(todo) ? accum : accum + 1;
  }, 0), [todos])

  return (
    <TodoPresentation
      inputRef={inputRef}
      onCreateTodo={onCreateTodo}
      showTodos={showTodos}
      onUpdateTodoStatus={onUpdateTodoStatus}
      deleteTodo={handleDeleteTodo}
      todos={todos}
      activeTodos={activeTodos}
      onToggleAllTodo={onToggleAllTodo}
      handleShowingAll={handleShowingAll}
      handleShowingActive={handleShowingActive}
      handleShowingCompleted={handleShowingCompleted}
      onDeleteAllTodo={onDeleteAllTodo}
      onUpdateTodoContent={handleUpdateTodoContent}
      showing={showing}
    />
  );
};

export default TodoComponent;