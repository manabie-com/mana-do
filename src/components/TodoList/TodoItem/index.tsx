import React, { useContext, useState } from "react";
import { Todo, TodoStatus } from "../../../models/todo";
import AppTag from "../../../commons/AppTag";
import AppText from "../../../commons/AppText";
import AppButton from "../../../commons/AppButton";
import AppInput from "../../../commons/AppInput";
import { AppContext } from "../../../store/context";
import { setTodos } from "../../../store/actions";
import { useHistory } from "react-router-dom";
import Service from "../../../service";
import AppCheckbox from "../../../commons/AppCheckbox";

interface ITodoItem {
    itemNo: number | string,
    todo: Todo,
}




const TodoItem = ({ todo, itemNo }: ITodoItem) => {
    const { dispatch } = useContext(AppContext);
    const [isEditing, setIsEditing] = useState(false);
    const [editingItem, setEditingItem] = useState(todo);
    const history = useHistory();

    const onSave = async (todo: Todo) => {
        try {
            const response = await Service.updateTodo(todo);
            if (response.status === 200) {
                dispatch( setTodos(response.data) )
            }
            setIsEditing(false);
        } catch (error) {
            if (error.status === 401) {
                history.push('/');
            }
        }
    }

    const onRemove = async (todo: Todo) => {
        try {
            const response = await Service.deleteTodo(todo);
            if (response.status === 200) {
                dispatch( setTodos(response.data) )
            }
        } catch (error) {
            if (error.status === 401) {
                history.push('/');
            }
        }
    }
    const handleOnClickEdit = (e:React.MouseEvent<HTMLElement>) => {
        e.preventDefault();
        setIsEditing(true);
    }

    const handleOnClickCancel = (e:React.MouseEvent<HTMLElement>) => {
        e.preventDefault();
        setIsEditing(false);
    }

    const onChangeField = (e:React.ChangeEvent<HTMLInputElement>) => {
        let changedValue = e.target.value;
        setEditingItem({...todo, content: changedValue});
    }

    const handleOnClickSave = (e:React.MouseEvent<HTMLElement>) => {
        onSave(editingItem);
    }
    const handleOnClickRemove = (e:React.MouseEvent<HTMLElement>) => {
        onRemove(todo);
    }

    const onKeyboardEnter = async (e: React.KeyboardEvent<HTMLInputElement>) => {
        if (e.key === 'Enter') {
            onSave(editingItem);
        }
    }

    const updateTodoStatus = async (todo:Todo,status: TodoStatus) => {
        try {
            const response = await Service.updateTodoStatus(todo, status);
            if (response.status === 200) {
                dispatch( setTodos(response.data) )
            }
        } catch (error) {
            if (error.status === 401) {
                history.push('/');
            }
        }
    }

    const onCheckboxChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        let value = e.target.checked;
        if (value) { 
            updateTodoStatus(todo, TodoStatus.COMPLETED) 
        } else {
            updateTodoStatus(todo, TodoStatus.ACTIVE)
        }
    }


    

    return (
        <tr>
            <td className="align-center">
                { itemNo}
            </td>
            <td className="align-center">
                <AppCheckbox defaultChecked={todo.status===TodoStatus.COMPLETED} onChange={(e: React.ChangeEvent<HTMLInputElement>)=>onCheckboxChange(e)}></AppCheckbox>
            </td>
            <td className="align-left">
                {!isEditing && <AppText strikethrough={todo.status===TodoStatus.COMPLETED}>{todo.content}</AppText>}
                {isEditing && <AppInput  className='edit-todo' onKeyDown={onKeyboardEnter} defaultValue={todo.content} onChange={(e: React.ChangeEvent<HTMLInputElement>)=>{onChangeField(e)}}></AppInput>}
            </td>
            <td className="align-center">
                <AppTag color={todo.status===TodoStatus.COMPLETED?"success":"warning"}>{todo.status}</AppTag>
            </td>
            <td className="align-center">
                {new Date(todo.created_date).toLocaleString('vi-VN')}
            </td>
            <td className="align-center">
                {!isEditing && <AppButton btnType="link" href="#" onClick={(e: React.MouseEvent<HTMLElement>)=>handleOnClickEdit(e)}>Edit</AppButton>}
                { isEditing && <AppButton btnType="link" href="#" onClick={(e: React.MouseEvent<HTMLElement>)=>handleOnClickSave(e)}>Save</AppButton>}
                { isEditing && <AppButton btnType="link" href="#" onClick={(e: React.MouseEvent<HTMLElement>)=>handleOnClickCancel(e)}>Cancel</AppButton>                 }
                <AppButton btnType="link" href="#" onClick={(e: React.MouseEvent<HTMLElement>)=>handleOnClickRemove(e)}>Remove</AppButton>
            </td>
        </tr>
    )
}

export default TodoItem;