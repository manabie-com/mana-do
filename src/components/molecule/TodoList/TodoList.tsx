import { Todo, TodoStatus } from "models/todo";
import React from "react";
import ToDoItem from "components/atom/TodoItem";

export type EnhanceTodoStatus = TodoStatus | "ALL";

type ToDoListProps = {
  showTodos: Todo[];
  showing: EnhanceTodoStatus;
};

export const emptyTodoMessage = (showing: EnhanceTodoStatus) =>
  `There are not${
    showing === TodoStatus.ACTIVE
      ? " active"
      : showing === TodoStatus.COMPLETED
      ? " completed"
      : ""
  } items`;

const ToDoList: React.FC<ToDoListProps> = ({ showTodos, showing }) => {
  return (
    <div className="ToDo__list">
      {showTodos.length === 0 && (
        <span className="Todo__empty__message" data-testid="empty-message">
          {emptyTodoMessage(showing)}
        </span>
      )}
      {showTodos.map((todo) => {
        return <ToDoItem todo={todo} key={todo.id} />;
      })}
    </div>
  );
};

export default ToDoList;
