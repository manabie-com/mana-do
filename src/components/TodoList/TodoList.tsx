import React from "react";

import { Todo } from "types";
import TodoItem from "components/TodoItem";

type PropTypes = {
  todos: Todo[];
};

export default function TodoList({ todos = [] }: PropTypes): JSX.Element {
  return (
    <div data-testid="todo-list">
      {todos.map(({ id, completed, content }) => (
        <TodoItem key={id} id={id} content={content} isCompleted={completed} />
      ))}
    </div>
  );
}
