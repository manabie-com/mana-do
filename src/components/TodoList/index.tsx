import React from "react";
import { Todo } from "models/todo";
import TodoItem from "components/TodoItem";

export interface TodoListProps {
  list: Todo[];
  onUpdateTodoStatus: (todoId: string, checked: boolean) => void;
  onDeleteTodo: (todoId: string) => void;
  onUpdateTodo: (todo: Todo) => void;
}

const TodoList: React.FC<TodoListProps> = ({
  list = [],
  onUpdateTodoStatus,
  onDeleteTodo,
  onUpdateTodo,
}) => {
  if (list.length === 0) {
    return (
      <div className="Todo__list--empty">You have nothing to do today.</div>
    );
  }
  return (
    <div className="ToDo__list">
      {list.map((todo, index) => (
        <TodoItem
          key={`todo-item-${index}`}
          todo={todo}
          onUpdateStatus={onUpdateTodoStatus}
          onDelete={onDeleteTodo}
          onUpdate={onUpdateTodo}
        />
      ))}
    </div>
  );
};

export default TodoList;
