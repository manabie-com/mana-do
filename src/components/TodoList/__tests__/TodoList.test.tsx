import React from "react";
import { expect, test, describe } from "@jest/globals";
import { render, screen } from "@testing-library/react";
import TodoList, { TodoListProps } from "../index";
import shortid from "shortid";
import { Todo, TodoStatus } from "models/todo";
import { createRenderer } from "react-test-renderer/shallow";

const mockedOnDelete = jest.fn();
const mockedOnUpdate = jest.fn();
const mockedOnUpdateStatus = jest.fn();

function renderTodosArray(
  todoContents: { content: string; status: TodoStatus }[]
): Todo[] {
  return todoContents.map((todo) => ({
    id: shortid(),
    user_id: "1",
    created_date: new Date().toISOString(),
    content: todo.content,
    status: todo.status,
  }));
}

describe("TodoList snapshot", () => {
  test("should render snapshot when todo list has no element", () => {
    const r = createRenderer();
    const props: TodoListProps = {
      list: [],
      onUpdateTodoStatus: mockedOnUpdateStatus,
      onDeleteTodo: mockedOnDelete,
      onUpdateTodo: mockedOnUpdate,
    };
    r.render(<TodoList {...props} />);
    expect(r.getRenderOutput()).toMatchSnapshot();
  });

  test("should render snapshot when todo list has 1 element", () => {
    const todos = renderTodosArray([
      { content: "Feed the dog", status: TodoStatus.ACTIVE },
    ]);
    const r = createRenderer();
    const props: TodoListProps = {
      list: todos,
      onUpdateTodoStatus: mockedOnUpdateStatus,
      onDeleteTodo: mockedOnDelete,
      onUpdateTodo: mockedOnUpdate,
    };
    r.render(<TodoList {...props} />);
    expect(r.getRenderOutput()).toMatchSnapshot();
  });
});

describe("TodoList functions", () => {
  test("should render 'You have nothing to do today.' sentence when todo list has no element.", () => {
    const props: TodoListProps = {
      list: [],
      onUpdateTodoStatus: mockedOnUpdateStatus,
      onDeleteTodo: mockedOnDelete,
      onUpdateTodo: mockedOnUpdate,
    };
    render(<TodoList {...props} />);
    const element = screen.getByText(
      "You have nothing to do today."
    ) as HTMLDivElement;
    expect(element).toBeDefined();
  });

  test("should render 1 element when todo list has 1 element.", () => {
    const todos = renderTodosArray([
      { content: "Feed the dog", status: TodoStatus.ACTIVE },
    ]);
    const props: TodoListProps = {
      list: todos,
      onUpdateTodoStatus: mockedOnUpdateStatus,
      onDeleteTodo: mockedOnDelete,
      onUpdateTodo: mockedOnUpdate,
    };
    render(<TodoList {...props} />);
    const elements = screen.getAllByTestId("todo-item");
    expect(elements.length).toBe(1);
  });
});
