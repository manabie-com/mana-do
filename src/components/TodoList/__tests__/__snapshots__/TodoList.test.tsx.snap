// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList snapshot should render snapshot when todo list has 1 element 1`] = `
<div
  className="ToDo__list"
>
  <TodoItem
    onDelete={[MockFunction]}
    onUpdate={[MockFunction]}
    onUpdateStatus={[MockFunction]}
    todo={
      Object {
        "content": "Feed the dog",
        "created_date": "2022-01-08T12:52:01.655Z",
        "id": "P-5aDjyEu",
        "status": "ACTIVE",
        "user_id": "1",
      }
    }
  />
</div>
`;

exports[`TodoList snapshot should render snapshot when todo list has no element 1`] = `
<div
  className="Todo__list--empty"
>
  You have nothing to do today.
</div>
`;
